climada/engine/impact.py

```diff
@@ -1736,6 +1737,72 @@ def stack_attribute(attr_name: str) -> np.ndarray:
+        Assign for each impact coordinate closest hazard coordinate.
+        -1 used for disatances > threshold in point distances. If raster hazard,
+        -1 used for centroids outside raster.
+
+        Parameters
+        ----------
+        hazard : Hazard
+            Hazard to match (with raster or vector centroids).
+        distance : str, optional
+            Distance to use in case of vector centroids.
+            Possible values are "euclidean", "haversine" and "approx".
+            Default: "euclidean"
+        threshold : float
+            If the distance (in km) to the nearest neighbor exceeds `threshold`,
+            the index `-1` is assigned.
+            Set `threshold` to 0, to disable nearest neighbor matching.
+            Default: 100 (km)
+        overwrite: bool
+            If True, overwrite centroids already present. If False, do
+            not assign new centroids. Default is True.
+
+        See Also
+        --------
+        climada.util.coordinates.assign_coordinates: method to associate centroids to
+            impact points
+        climada.util.coordinates.assign_gdf_centroids: method to associate centroids to
+            geodataframe points
+
+        Notes
+        -----
+        The default order of use is:
+            1. if centroid raster is defined, assign impact points to
+            the closest raster point.
+            2. if no raster, assign centroids to the nearest neighbor using
+            euclidian metric.
+        Both cases can introduce inaccuracies for coordinates in lat/lon
+        as distances in degrees differ from distances in meters on the Earth
+        surface, in particular for higher latitude and distances larger than 100km.
+        If more accuracy is needed, please use 'haversine' distance metric.
+        This however is slower for (quasi-)gridded data, and works only for
+        non-gridded data.
+        """
+
+        haz_type = hazard.tag.haz_type
+        centr_haz = 'centr_' + haz_type
+
+        #Assert that the imp crs is epsg:4326, as it is required by the u_coord methods
+        if not u_coord.equal_crs(self.crs, 'EPSG:4326'):
+            raise ValueError('Set Impact to lat/lon crs (EPSG:4326)!')
+
+        #create geodataframe from coordinates
+        coord_df = pd.DataFrame(self.coord_exp, columns=['latitude', 'longitude'])
+        geometry = gpd.points_from_xy(coord_df.longitude,coord_df.latitude,crs = self.crs)
+        coord_gdf = gpd.GeoDataFrame(coord_df,geometry = geometry)
+
+        #call the assign_gdf_centroids util function
+        coord_gdf[centr_haz] = u_coord.assign_gdf_centroids(coord_gdf, hazard.centroids,
+                                                distance=distance,
+                                                threshold=threshold)
+
+        #create new impact attribute with hazard centroids
+        setattr(self,f'centr_{hazard.tag.haz_type}' , coord_df[f'centr_{hazard.tag.haz_type}'])
```

Comment:
Introducing an attribute is cumbersome to some extent. There are read/write methods to be considered and tested,. All methods that use impact as argument may in theory be affected, etc, etc,.
With exposures there is nothing we can do about it, the damage is done, but with impact we still have a chance to stay side-effect free. 🥺

Reply:
For our use case (impact function calibration with impact objects that store the recorded damages) it was very convenient to have the assigned centroids stored within the impact object. Then we don't always have to pass the assigned centroids seperately to other functions. 

If it is strongly discouraged, we could also just let the method return the ``coord_gdf`` and we would just manually set it as an attribute of the impact object. Would this be preferred?

Reply:
Yes, sorry, but this would actually be preferred.
Generally, in Python adjusting data types (classes) is very easy and seems to come in handy at many occasions. However the ease is much more of a curse than a blessing.
In this particular case it's most likely even justified to store the centroids assignment within the impact class, but nevertheless: patching a class with an attribute on the fly is not best practice, strictly speaking, and would require a thorough analysis of the introduced side effect and, as stated, _lots_ of tests.

Reply:
That's fine it can easily be changed. And I'm happy to learn from you regarding python best practices ;)

Reply:
🤓